### ISSUE TITLE
DeepChem Dependency Management

### ISSUE THREAD

**Issue by ncfrey** [deepchem/deepchem#2504]
## ❓ Question and Discussion

Package management is probably one of the single biggest headaches in Python, and DeepChem has inherited this issue. As more features and dependencies are added, package versions become harder to reconcile and keep up with. Here we can discuss some possible ways forward. 

Two that come to my mind are:
1. Revitalize the repo structure of the [deepchem organization](https://github.com/deepchem) and split packages into DeepChem, DeepMat, DeepBio, etc. As the "central science", DeepChem could set the API conventions and parent classes, but dependencies would be split among different packages rather than centrally maintained in a monorepo. Pros are that each package will have minimal dependencies and maximal flexibility to work with other packages. Cons are that it will be difficult to maintain multiple different packages and this would introduce a major API redesign.
2. Turn DeepChem into a fully-fledged "app" that is containerized with [Pipfile Lock](https://realpython.com/pipenv-guide/#the-pipfilelock), [Conda lock](https://pypi.org/project/conda-lock/), and/or Docker. Pros are that anyone will be able to run DeepChem anywhere, and we only have to maintain a single set of working package versions. Cons are that users who wish to use additional packages that are not direct dependencies will have to find versions that are compatible with the DC environment, or else split work across different environments.

I've added these ideas to get the discussion going. Please add other design ideas or comments on these!

**Comment by rbharath** [comment]
Thanks for kicking off this discussion! Here are some more thoughts (somewhat scattered for now, but putting up to raise discussion)

1. One variant of this idea is we could maintain multiple packages within the deepchem monorepo (we could have separate docs pages / pip packages but all packaged inside the deepchem/ monorepo for development ease). If we choose this path, it would have to be done gradually over several releases to maintain strong backwards compatibility for users which would introduce planning challenges. An advantage is that this may help us pull in new users into the community (a fluids simulation person might sensibly not realize DeepChem could be a useful tool for them right now!)
2. When I use DeepChem for applications, it's often my "primary dependency." Since DeepChem comes batteries included, usually I only end up using a few auxiliary extra packages (say visualization tools or other lightweight things). Do other people also use DeepChem this way? Or do you often use other heavy duty tools in conjunction? (A prominent counterexample is huggingface+deepchem is used by a number of folks but not quite supported fully by the main DeepChem yet).

**Comment by ncfrey** [comment]
> 2. When I use DeepChem for applications, it's often my "primary dependency." Since DeepChem comes batteries included, usually I only end up using a few auxiliary extra packages (say visualization tools or other lightweight things). Do other people also use DeepChem this way? Or do you often use other heavy duty tools in conjunction? (A prominent counterexample is huggingface+deepchem is used by a number of folks but not quite supported fully by the main DeepChem yet).

I strongly agree with this point. I use conda and pipenv environments to manage everything, because python package management is so tricky. I have a DeepChem environment, just like I have a Pymatgen/Matminer environment. I think any python project that becomes sufficiently powerful/complex requires this kind of approach. I'd be curious to see how others are using DeepChem.

**Comment by peastman** [comment]
> One variant of this idea is we could maintain multiple packages within the deepchem monorepo (we could have separate docs pages / pip packages but all packaged inside the deepchem/ monorepo for development ease).

This seems to me like a good approach.  All the code can stay in a single repository, but get split into multiple conda/pip packages.  People will install only the parts they need, bringing in only the dependencies they need.

> Since DeepChem comes batteries included, usually I only end up using a few auxiliary extra packages (say visualization tools or other lightweight things).

I think this is contradicted by the large and steadily growing set of dependencies.  Most of them started out as unrelated packages that someone wanted to use with DeepChem.  Then they contributed some code to DeepChem that required it, and instantly it became a dependency.  This isn't slowing down, because people keep finding other packages they want to use with DeepChem.  It isn't sustainable.  It's already difficult to find consistent versions of all the dependencies, and each added dependency makes it harder.  Eventually it will become impossible.

A better way of approaching it is to say that matminer, mordred, simdna, networkx, pdbfixer, pubchempy, etc. are all examples of independent packages that certain people want to use with DeepChem.  That doesn't mean they're dependencies, and you shouldn't need to install them to use DeepChem.

**Comment by rbharath** [comment]
> This seems to me like a good approach. All the code can stay in a single repository, but get split into multiple conda/pip packages. People will install only the parts they need, bringing in only the dependencies they need.

Here's a quick sketch for what a DeepChem future package structure could look like:

- deepscience: The core DeepChem modules and classes which are application independent (datasets, generic splitters/transformers/models). deepscience should have very minimal dependencies restricted to core python libraries (numpy/pandas, maybe tf/pytorch)
  - deepchem: Chemistry specific featurizers/splitters/models. To prevent large user breakage, we'd want to import core modules from deepscience so existing deepchem code bases could continue working as expected
  - deepmat: Materials science specific featurizers/splitters/models etc
  - deepbio: Protein sequence featurizers/splitters/models etc
  - More modules can be added as we expand into new domains

> I think this is contradicted by the large and steadily growing set of dependencies. Most of them started out as unrelated packages that someone wanted to use with DeepChem. Then they contributed some code to DeepChem that required it, and instantly it became a dependency. This isn't slowing down, because people keep finding other packages they want to use with DeepChem. It isn't sustainable. It's already difficult to find consistent versions of all the dependencies, and each added dependency makes it harder. Eventually it will become impossible.
> 
> A better way of approaching it is to say that matminer, mordred, simdna, networkx, pdbfixer, pubchempy, etc. are all examples of independent packages that certain people want to use with DeepChem. That doesn't mean they're dependencies, and you shouldn't need to install them to use DeepChem.

This is an excellent counterpoint. Following a split like the one I've outlined above may help ameliorate the problem of ballooning dependencies

**Comment by peastman** [comment]
Another possibility is for everything to remain "DeepChem", but split the installation into multiple packages.  A model for this is OpenAI Gym.  If you just `pip install gym`, that gets you the core library and a minimal set of environments.  There are lots of other environments, many of which require additional dependencies, so they split them into additional packages that you install separately.

**Comment by peastman** [comment]
A more extreme version of that would be for the secondary packages to provide *only* dependencies.  If you installed just the core DeepChem package, you would get the full source code but only a few essential dependencies like scikit-learn and TensorFlow.  That would mean a lot of features would throw exceptions if you tried to use them.  Then there could be a `deepchem-pytorch` package that would install PyTorch, DGL, Pytorch Geometric, etc.  And there could be a `deepchem-molecules` package that would install RDKit, MDTraj, PDBFixer, etc.  And so on.

This probably would only be practical for packages installed with conda, but that's going to be an issue regardless.

**Comment by rbharath** [comment]
I'm on-board with the strategy of keeping everything as "DeepChem" and splitting up into multiple installation packages (which are possibly just dependencies). The larger deepscience/deepchem/deepmat split would be a lot more work and picking the simpler strategy to start seems smart.

Looking at the openai gym setup.py file:

https://github.com/openai/gym/blob/master/setup.py

It looks like this could be relatively straightforward to do? We'd need to modify `setup.py` and then modify the github actions release infrastructure to release multiple package

**Comment by peastman** [comment]
setup.py already lists only a minimal set of dependencies:

https://github.com/deepchem/deepchem/blob/875f7f465a3e3d285a0364c2519a3f2a1407a47b/setup.py#L65-L71

**Comment by rbharath** [comment]
Closing this old issue since we've mostly stabilized on the idea of one large deepchem package but trying to keep the dependencies as small as possible. We still have major CI issues, but I will raise another issue to discuss those
